### PR TITLE
refactor(client): scope added in memory

### LINKED ISSUE #431: return granted scopes used in client-tokens list
We should show the user the scopes that were used when they logged in into some relier.
We should also show those somehow in the Devices and Apps list.

### PR DESCRIPTION
**Pull request by divyabiyani**
Fixes #431.

### CODE REVIEW

File: lib/db/memory.js [hunk 1]
```diff
@@ -269,7 +270,14 @@ MemoryStore.prototype = {
       if (client.canGrant === false) {
         if (! activeClients[clientIdHex]) {
           activeClients[clientIdHex] = client;
+          activeClients[clientIdHex].scope = [];
         }
+        clientTokenObj.clientScope.forEach(function (scope) {
```
Comment (vladikoff) [inline lib/db/memory.js]: @divyabiyani This needs to handle this transformation:

with 3 token scopes of 1) `clients:write` 2) `profile profile:write` 3) `profile` you need to end up with: 

`clients:write profile profile:write`. 

Would this code here become:  `clients:write profile profile:write profile` ?

Comment (divyabiyani) [inline lib/db/memory.js]: No, @vladikoff. It is giving scope="clients:write,profile,profile:write"

Comment (vladikoff) [inline lib/db/memory.js]: Oh sorry I got confused with the way scopes are stored. Let's add a test for this new behaviour. 
We need the test to issue ~3-4 tokens with different scopes and return the unified list.

File: lib/db/mysql/index.js [hunk 2]
```diff
@@ -431,7 +430,67 @@ MysqlStore.prototype = {
...
+          }
+        });
+
+        var clientTokenTime = clientTokenObj.createdAt;
+        if (clientTokenTime > activeClients[clientIdHex].createdAt) {
+          activeClients[clientIdHex].createdAt = clientTokenTime;
+        }
+      });
+      var activeClientsArray = Object.keys(activeClients).map(function (key) {
+        var Scopes = activeClients[key].scope;
```
Comment (vladikoff) [inline lib/db/mysql/index.js]: no need to make `Scopes`, no need for the capital `S` in there...

File: lib/db/mysql/index.js [hunk 3]
```diff
@@ -431,7 +430,67 @@ MysqlStore.prototype = {
   getActiveClientTokensByUid: function getActiveClientTokensByUid(uid) {
     return this._read(QUERY_ACTIVE_CLIENT_TOKENS_BY_UID, [
       buf(uid)
-    ]);
+    ]).then(function(activeClientIds) {
+      var activeClients = {};
```
Comment (vladikoff) [inline lib/db/mysql/index.js]: do you think we can make a single function that will work for both (memory and MySQL engines)?

File: test/api.js [hunk 4]
```diff
@@ -2286,13 +2286,13 @@ describe('/v1', function() {
         return getUniqueUserAndToken(client1Id.toString('hex'), {
           uid: user1.uid,
           email: user1.email,
-          scopes: ['profile']
+          scopes: ['profile', 'profile:write']
```
Comment (vladikoff) [inline test/api.js]: @divyabiyani could we please add a new test just for this with more advanced cases, such as:

Same client:

1) scopes: ['profile', 'profile:write']
2) scopes: ['clients:write']
3) scopes: ['basket:write', 'profile:email']
4) scopes: ['profile:uid', 'profile', 'profile:write']

In this new test we just care about scopes, so we don't need the extra assertions.

Comment (divyabiyani) [inline test/api.js]: Ok @vladikoff . I will add it.

File: lib/db/mysql/index.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ const MysqlPatcher = require('mysql-patcher');
 const config = require('../../config');
 const encrypt = require('../../encrypt');
 const P = require('../../promise');
+const scope = require('../scope');
 const Scope = require('../../scope');
```
Comment (vladikoff) [inline lib/db/mysql/index.js]: @divyabiyani hm the `scope` and `Scope` would be really confusing, we need to rename of them

File: lib/db/scope.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const unbuf = require('buf').unbuf.hex;
+
+const Scope = {
```
Comment (vladikoff) [inline lib/db/scope.js]: @divyabiyani does it make sense to call this `helpers` or something? and rename this to `lib/db/helpers.js`?

We also need a unit test for every function in this module

Comment (vladikoff) [inline lib/db/scope.js]: you can probably just do `module.exports = {
// helpers functions and code here
}

File: lib/db/scope.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const unbuf = require('buf').unbuf.hex;
```
Comment (vladikoff) [inline lib/db/scope.js]: let's add a comment above here to explain what this file is for. 
Example: `This module consists of helpers that are used by both memory and MySQL database engines`

**Comment by vladikoff** [comment]
Continued in https://github.com/mozilla/fxa-oauth-server/pull/445

**Pull request closed without merging**